app/services/feeds/import.rb

```diff
@@ -0,0 +1,152 @@
+# TODO: [rhymes]
+# => add Feeds::ImportUser to fetch a single user
+# => add Feeds::ValidateFeedUrl to validate a single feed URL
+module Feeds
+  class Import
+    def self.call(silence: true)
+      # TODO: remove this eventually
+      if Rails.env.development? && silence
+        ActiveRecord::Base.logger = nil
+        Rails.logger.level = :info
+        Rails.configuration.log_level = :info
+      end
+
+      new.call
+    end
+
+    # TODO: add `users` param
+    def initialize
+      @users = User.where.not(feed_url: [nil, ""])
+
+      # NOTE: should these be configurable? Currently they are the result of empiric
+      # tests trying to find a balance between memory occupation and speed
+      @users_batch_size = 50
+      @num_fetchers = 8
+      @num_parsers = 4
+    end
+
+    def call
+      total_articles_count = 0
+
+      users.in_batches(of: users_batch_size) do |batch_of_users|
+        feeds_per_user_id = fetch_feeds(batch_of_users)
+        Rails.logger.info("feeds_per_user_id.length: #{feeds_per_user_id.length}")
+
+        feedjira_objects = parse_feeds(feeds_per_user_id)
+        Rails.logger.info("feedjira_objects.length: #{feedjira_objects.length}")
+
+        # NOTE: doing this sequentially to avoid locking problems with the DB
+        # and unnecessary conflicts
+        articles = feedjira_objects.map do |user_id, feed|
+          # TODO: replace `feed` with `feed.url` as `RssReader::Assembler`
+          # only actually needs feed.url
+          user = batch_of_users.detect { |u| u.id == user_id }
+          create_articles_from_user_feed(user, feed)
+        end.flatten
+        Rails.logger.info("articles.length: #{articles.length}")
+
+        total_articles_count += articles.length
+
+        articles.each { |article| Slack::Messengers::ArticleFetchedFeed.call(article: article) }
+      end
+
+      total_articles_count
+    end
+
+    private
+
+    attr_reader :users, :users_batch_size, :num_fetchers, :num_parsers
+
+    # TODO: put this in separate service object
+    def fetch_feeds(batch_of_users)
+      data = batch_of_users.pluck(:id, :feed_url)
+
+      result = Parallel.map(data, in_threads: num_fetchers) do |user_id, url|
+        response = HTTParty.get(url.strip, timeout: 10)
+
+        [user_id, response.body]
+      rescue StandardError
+        # TODO: add exception handling
+        # For example, we should stop pulling feeds that return 404 and disable them?
+        nil
+      end
+
+      batch_of_users.update_all(feed_fetched_at: Time.current)
+
+      Hash[result.compact]
+    end
+
+    # TODO: put this in separate service object
+    def parse_feeds(feeds_per_user_id)
+      result = Parallel.map(feeds_per_user_id, in_threads: num_parsers) do |user_id, feed_xml|
```

Comment:
Let's keep a close eye on this, I don't want a process getting too large since we are already throwing multiple threads inside an already threaded process.

Reply:
Yes, this is exactly the kind of thing we need to monitor :)